### PR TITLE
#839 Improve revoked Google access handling

### LINKED ISSUE #839: better handling of revoked google access
1) set up the app
2) go to https://myaccount.google.com/permissions and remove access from iOS app
3) go back to the app

actual result: ugly error with retries

expected result: `Please reconnect the app to Gmail. [Continue]` where the button will allow user to go through sign-in again (but not the rest of setup) and save new sign-in tokens. After that, send the user back to where the error just happened, and it should work again.

### PR DESCRIPTION
**Pull request by sosnovsky**
This PR adds error screen when user revoked Google access and needs to re-authorize app
close #839

<image src="https://user-images.githubusercontent.com/1062093/166473179-f97f99b0-231e-498f-a93c-9cfe5fb4310e.png" width="300">

----------------------------------

**Tests** _(delete all except exactly one)_:
- Difficult to test (explain why) - revoking Google access is complicated in test environment

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by sosnovsky** [comment]
@tomholub I have only https://github.com/FlowCrypt/flowcrypt-ios/issues/1443 left assigned to me, can you please add more tasks? Thanks

**Comment by ioanmo226** [comment]
@sosnovsky Maybe better if we add `Logout` button in this screen?
In case user has multiple accounts setup and don't want to reauthorize?

image

**Comment by tomholub** [comment]
@ioanmo226 you can get this PR tested / approved / merged between you two once you are both happy with it.

**Comment by sosnovsky** [comment]
> @sosnovsky Maybe better if we add `Logout` button (and also revoked access user's email) in this screen? In case user has multiple accounts setup and don't want to reauthorize?

Great idea, I'll add `Logout` there

**Comment by ioanmo226** [comment]
How about adding user's email address in textfield? Something like `X account google account access was revoked. Please reconnect the app to Gmail`?
User might have multiple accounts setup and he might wants to know which account access was revoked?

**Comment by sosnovsky** [comment]
@ioanmo226 I updated 'invalid grant' screen to contain revoked email and `logout` button

**Comment by ioanmo226** [comment]
Nice. That looks great

File: FlowCrypt/Functionality/Services/GlobalRouter.swift [hunk 1]
```diff
@@ -177,22 +189,24 @@ extension GlobalRouter: GlobalRouterType {
...
-                let checkAuthViewController = CheckMailAuthViewController(appContext: appContext)
+                let rootViewController = keyWindow.rootViewController
+                let navigationController = rootViewController as? UINavigationController ?? rootViewController?.navigationController
+                let checkAuthViewController = CheckMailAuthViewController(
+                    appContext: appContext,
+                    decorator: CheckMailAuthViewDecorator(type: .setup)
+                )
                 navigationController?.pushViewController(checkAuthViewController, animated: true)
                 return
             }
```
[the remark this answers is not in the document]
Comment (sosnovsky) [reply FlowCrypt/Functionality/Services/GlobalRouter.swift:204]: @ioanmo226 We need `return` here as app shouldn't show error modal only when user cancels login, while still showing it for other errors
```swift
if case .cancelledAuthorization = gmailUserError {
  return
}
```

Adding `return` on line 205 will disable error modals for all `GoogleUserServiceError`, but that's not what we need.
I can add additional comment to the code, as these logic can be difficult to understand.

[the remark this answers is not in the document]
Comment (ioanmo226) [reply FlowCrypt/Functionality/Services/GlobalRouter.swift:204]: Sounds good.

File: FlowCrypt/Functionality/Services/GlobalRouter.swift [hunk 2]
```diff
@@ -55,20 +56,21 @@ extension GlobalRouter: GlobalRouterType {
...
+    func signIn(appContext: AppContext, route: GlobalRoutingType, email: String?) async {
         logger.logInfo("Sign in with \(route)")
         do {
             switch route {
             case .gmailLogin(let viewController):
                 viewController.showSpinner()
 
                 let googleService = GoogleUserService(
-                    currentUserEmail: try appContext.encryptedStorage.activeUser?.email,
+                    currentUserEmail: email,
```
[the remark this answers is not in the document]
Comment (sosnovsky) [reply FlowCrypt/Functionality/Services/GlobalRouter.swift:67]: App passes `nil` here in `SignInViewController` - https://github.com/FlowCrypt/flowcrypt-ios/blob/bugfix/issue-839-google-access-handling/FlowCrypt/Controllers/SignIn/SignInViewController.swift#L117 (it's used when user tries to add second account in the app)

We pass `email` when user needs to re-authorize, so his email is already stored in the app.

[the remark this answers is not in the document]
Comment (sosnovsky) [reply FlowCrypt/Functionality/Services/GlobalRouter.swift:67]: When adding second account we pass current `appContext` (which contains first account email) to `SignInViewController` https://github.com/FlowCrypt/flowcrypt-ios/blob/master/FlowCrypt/Controllers/SideMenu/Menu/MyMenuViewController.swift#L221

Then tapping on `Sign in with Google` tries to sign in to existing account and makes it impossible to add another one.

[the remark this answers is not in the document]
Comment (ioanmo226) [reply FlowCrypt/Functionality/Services/GlobalRouter.swift:67]: Got it. +1
Thanks for your info.

**Comment by ioanmo226** [comment]
Everything looks okay, Please add comment as we agreed (for GoogleUserServiceError ) and I'll approve this PR.

**Comment by sosnovsky** [comment]
> Everything looks okay, Please add comment as we agreed (for GoogleUserServiceError ) and I'll approve this PR.

Done ✅

**Comment by ioanmo226** [comment]
@sosnovsky You may enable auto-merge as I don't have permission for that.

**Comment by ioanmo226** [comment]
@sosnovsky I think next priority issue should be #1568. CheckRefreshKeyFromEkmErrorHandling ui test sometimes fails (it fails 1-2 times for 10 times) in semaphoreci.

**Comment by sosnovsky** [comment]
> @sosnovsky I think next priority issue should be #1568. CheckRefreshKeyFromEkmErrorHandling ui test sometimes fails (it fails 1-2 times for 10 times) in semaphoreci.

it's from the next milestone `1.0.3`, while I have several tasks in the current `1.0.2` milestone, so probably it's not so critical.

**Comment by ioanmo226** [comment]
ok

**Comment by tomholub** [comment]
I have bumped that issue to 1.0.2. If it's affecting development, it would be good to get it resolved

**Pull request merged**